### DISCUSSION TITLE
FVM M2.2: Replace the Constructor method with a wasm-level "constructor" entrypoint

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#500] [Enhancements - Technical]
For some background, Filecoin "sends a message" to an actor (i.e., "calls" an FVM actor) by calling a single wasm-level `invoke` function. This `invoke` function internally dispatches on the method number.

When a user _constructs_ a new actor:

1. The user's actor/account calls the `Exec` method on the init actor.
2. The init actor allocates a new actor ID/address, and records it in its address map.
3. The init actor creates a "stub" actor with empty state.
4. The init actor calls new actor's `Constructor` method (method 1) using the normal calling convention. Internally, this will call the `invoke` function on the actor's wasm module.

However, this has two significant drawbacks:

1. The new actor's constructor sees the _init_ actor as the caller, not the user's actor.
2. The constructor has to _carefully_ check that the caller _is_ the init actor. Technically, the constructor can be called by anyone at runtime (although every sane actor will just reject the message).

I'd like to propose an alternative: introduce a new wasm-level function called `constructor`. In this new world:

1. The user's actor/account calls the `Exec` method on the init actor.
2. The init actor allocates a new actor ID/address, and records it in its address map.
3. The init actor asks the FVM (via a privileged syscall) to construct a new actor at the newly allocated address. 
4. The FVM calls the `constructor` function on the actor's WASM module, exporting the user's actor from step 1 as the "caller".

**Alternatives**

An alternative is to:

1. Make "method 1" privileged and prevent anyone from sending a method 1 message except the init actor.
2. Allow the init actor to "send as", or somehow expose the ultimate "creator's" address to the newly created actor.

**Comment by anorth** [comment]
Ack, no objections from me to this basic idea.
